File: backend/src/main/java/com/ll/nbe342team8/domain/book/book/dto/BookResponseDto.java
package com.ll.nbe342team8.domain.book.book.dto;

import java.time.LocalDate;

import com.ll.nbe342team8.domain.book.book.entity.Book;
import com.ll.nbe342team8.domain.book.category.entity.Category;

public record BookResponseDto(
		Long id,
		String title,
		String author,
		String isbn,
		String isbn13,
		String publisher,
		LocalDate pubDate,
		int priceStandard,
		int priceSales,
		Long salesPoint,
		int stock,
		int status,
		Double rating,
		String toc,
		Long reviewCount,
		String coverImage,
		Integer categoryId,
		String description,
		String descriptionImage,
		Double averageRating  // 추가한 평균 평점 필드
) {
	public static BookResponseDto from(Book book) {
		return new BookResponseDto(
				book.getId(),
				book.getTitle(),
				book.getAuthor(),
				book.getIsbn(),
				book.getIsbn13(),
				book.getPublisher(),
				book.getPubDate(),
				book.getPriceStandard(),
				book.getPricesSales(),
				book.getSalesPoint(),
				book.getStock(),
				book.getStatus(),
				book.getRating(),
				book.getToc(),
				book.getReviewCount(),
				book.getCoverImage(),
				book.getCategoryId().getCategoryId(),
				book.getDescription(),
				book.getDescriptionImage(),
				book.getAverageRating()
		);
	}

	// 외부 API ExternalBookDto -> 내부 DTO 변환
	// 여기서는 @NotNull인 필드에 대해 null 체크 후 기본값 적용
	public static BookResponseDto from(ExternalBookDto externalBookDto) {
		return new BookResponseDto(
				null, // id는 DB 저장 시 자동 생성
				externalBookDto.title() != null ? externalBookDto.title() : "", // 필수: 제목
				externalBookDto.author() != null ? externalBookDto.author() : "", // 필수: 저자
				externalBookDto.isbn() != null ? externalBookDto.isbn() : "0000000000000", // 선택: ISBN
				externalBookDto.isbn13() != null ? externalBookDto.isbn13() : "9999999999999", // 필수: ISBN13
				externalBookDto.publisher() != null ? externalBookDto.publisher() : "Unknown", // 선택: 출판사
				externalBookDto.pubDate() != null ? externalBookDto.pubDate() : LocalDate.of(9999, 12, 31), // 필수: 출판일
				externalBookDto.priceStandard() != null ? externalBookDto.priceStandard() : 9999999, // 필수: 정가
				externalBookDto.priceSales() != null ? externalBookDto.priceSales() : 9999999, // 필수: 판매가
				0L, // 필수: 판매 포인트 (기본값 0L)
				0, // 필수: 재고 (외부 데이터에 없으므로 0 기본)
				0, // 필수: 판매 상태 (외부 데이터에 없으므로 0 기본, 품절)
				0.0, // 선택: 평점 (없으면 0.0 기본값)
				externalBookDto.toc() != null ? externalBookDto.toc() : "", // 선택: 목차
				0L, // 선택: 리뷰 수 (없으면 0L 기본값)
				externalBookDto.cover() != null ? externalBookDto.cover() : "", // 선택: 커버 이미지
				externalBookDto.categoryId() != null ? externalBookDto.categoryId() : 99999, // 필수: 카테고리 ID (없으면 99999로 기본)
				externalBookDto.description() != null ? externalBookDto.description() : "", // 선택: 상세 설명
				externalBookDto.descriptionImage() != null ? externalBookDto.descriptionImage() : "", // 선택: 상세 설명 이미지
				0.0 // 선택: 평균 평점 (없으면 0.0 기본값)
		);
	}


	// BookResponseDto -> Book 엔티티 변환
	public Book toEntity(Category category) {
		return Book.builder()
				.title(this.title != null ? this.title : "")                                 // 필수: 제목
				.author(this.author != null ? this.author : "")                              // 필수: 저자
				.publisher(this.publisher)                                                   // 선택: 출판사
				.isbn(this.isbn)                                                             // 선택: ISBN
				.isbn13(this.isbn13 != null ? this.isbn13 : "9999999999999")                 // 필수: ISBN13
				.pubDate(this.pubDate != null ? this.pubDate : LocalDate.of(9999, 12, 31))   // 필수: 출판일
				.priceStandard(this.priceStandard)                                           // 필수: 정가
				.pricesSales(this.priceSales)                                               // 필수: 판매가
				.stock(this.stock)                                                           // 필수: 재고
				.status(this.status)                                                         // 필수: 상태
				.salesPoint(this.salesPoint != null ? this.salesPoint : 0L)                  // 선택: 판매량
				.reviewCount(this.reviewCount != null ? this.reviewCount : 0L)               // 선택: 리뷰 수
				.rating(this.rating != null ? this.rating : 0.0)                             // 선택: 평점
				.averageRating(this.averageRating != null ? this.averageRating : 0.0)        // 선택: 평균 평점
				.toc(this.toc != null ? this.toc : "")                                       // 선택: 목차
				.coverImage(this.coverImage != null ? this.coverImage : "")                  // 선택: 커버 이미지
				.description(this.description != null ? this.description : "")               // 선택: 상세 설명
				.descriptionImage(this.descriptionImage != null ? this.descriptionImage : "")// 선택: 상세 설명 이미지
				.categoryId(category)                                                          // 필수: 카테고리 (외부에서 제공해야 함)
				.build();
	}
}

File: backend/src/main/java/com/ll/nbe342team8/domain/book/book/dto/BookResponseDto.kt
package com.ll.nbe342team8.domain.book.book.dto

import com.ll.nbe342team8.domain.book.book.entity.Book
import com.ll.nbe342team8.domain.book.category.entity.Category
import java.time.LocalDate

data class BookResponseDto(
    val id: Long?,
    val title: String,
    val author: String,
    val isbn: String?,
    val isbn13: String,
    val publisher: String?,
    val pubDate: LocalDate,
    val priceStandard: Int,
    val priceSales: Int,
    val salesPoint: Long,
    val stock: Int,
    val status: Int,
    val rating: Double,
    val toc: String,
    val reviewCount: Long,
    val coverImage: String,
    val categoryId: Int,
    val description: String,
    val descriptionImage: String,
    val averageRating: Double
) {
    companion object {
        fun from(book: Book): BookResponseDto {
            return BookResponseDto(
                id = book.id,
                title = book.title,
                author = book.author,
                isbn = book.isbn,
                isbn13 = book.isbn13,
                publisher = book.publisher,
                pubDate = book.pubDate,
                priceStandard = book.priceStandard,
                priceSales = book.pricesSales,
                salesPoint = book.salesPoint ?: 0L,
                stock = book.stock,
                status = book.status,
                rating = book.rating,
                toc = book.toc ?: "",
                reviewCount = book.reviewCount,
                coverImage = book.coverImage,
                categoryId = book.categoryId.categoryId,
                description = book.description ?: "",
                descriptionImage = book.descriptionImage ?: "",
                averageRating = book.averageRating ?: 0.0
            )
        }

        fun from(externalBookDto: ExternalBookDto): BookResponseDto {
            return BookResponseDto(
                id = null, // DB 저장 시 자동 생성
                title = externalBookDto.title ?: "",
                author = externalBookDto.author ?: "",
                isbn = externalBookDto.isbn ?: "0000000000000",
                isbn13 = externalBookDto.isbn13 ?: "9999999999999",
                publisher = externalBookDto.publisher ?: "Unknown",
                pubDate = externalBookDto.pubDate ?: LocalDate.of(9999, 12, 31),
                priceStandard = externalBookDto.priceStandard ?: 9999999,
                priceSales = externalBookDto.priceSales ?: 9999999,
                salesPoint = 0L,
                stock = 0,
                status = 0,
                rating = 0.0,
                toc = externalBookDto.toc ?: "",
                reviewCount = 0L,
                coverImage = externalBookDto.cover ?: "",
                categoryId = externalBookDto.categoryId ?: 99999,
                description = externalBookDto.description ?: "",
                descriptionImage = externalBookDto.descriptionImage ?: "",
                averageRating = 0.0
            )
        }
    }

    // BookResponseDto -> Book 엔티티 변환 (Category는 외부에서 제공)
    fun toEntity(category: Category): Book {
        return Book.builder()
            .title(if (this.title.isNotEmpty()) this.title else "")
            .author(if (this.author.isNotEmpty()) this.author else "")
            .publisher(this.publisher)
            .isbn(this.isbn)
            .isbn13(if (this.isbn13.isNotEmpty()) this.isbn13 else "9999999999999")
            .pubDate(this.pubDate)
            .priceStandard(this.priceStandard)
            .pricesSales(this.priceSales)
            .stock(this.stock)
            .status(this.status)
            .salesPoint(this.salesPoint)
            .reviewCount(this.reviewCount)
            .rating(this.rating)
            .averageRating(this.averageRating)
            .toc(this.toc.ifEmpty { "" })
            .coverImage(this.coverImage.ifEmpty { "" })
            .description(this.description.ifEmpty { "" })
            .descriptionImage(this.descriptionImage.ifEmpty { "" })
            .categoryId(category)
            .build()
    }
}